### PR TITLE
Add GetBlobMetaData and SetBlobMetaData  procedure to ABS Blob Client codeunit

### LINKED ISSUE #976: [Request] Add GetBlobMetadata procedure to ABS Blob Client codeunit
Hello,

In codeunit 9053 ABS Blob Client there is a procedure with several overloads GetBlobTags that allows us to get the tags associated with a particular blob in the Azure Blob Storage container. The tags can be retrieved as either an XmlDocument or Dictionary[Text, Text].

Is it possible to add similar procedures GetBlobMetadata to allow us to also get the metadata associated with a particular blob using ABS Blob Client?

Thanks in advance for your time.

### PR DESCRIPTION
**Pull request by Indubhusan23**
#### Summary 
In codeunit 9053 ABS Blob Client there is a procedure with several overloads GetBlobTags that allows us to get the tags associated with a particular blob in the Azure Blob Storage container. The tags can be retrieved as either an XmlDocument or Dictionary[Text, Text].

I added similar procedures GetBlobMetaData and SetBlobMetaData to allow us to also Get/Set the metadata associated with a particular blob using ABS Blob Client.

#### Work Item(s) 
Fixes #976

Fixes [AB#532988](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/532988)

### CODE REVIEW

**Comment by Indubhusan23** [comment]
@microsoft-github-policy-service agree [company="Zig"]

**Comment by Indubhusan23** [comment]
@microsoft-github-policy-service agree company="Zig"

**Comment by Indubhusan23** [comment]
> @Indubhusan23 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="Zig"

**Comment by JesperSchulz** [comment]
Moving the issue from ALAppExtensions to BCApps for triage.

File: src/System Application/App/Azure Blob Services API/src/ABSOperationResponse.Codeunit.al [hunk 1]
```diff
@@ -93,6 +93,25 @@ codeunit 9050 "ABS Operation Response"
         exit(Values[1]);
     end;
 
+    [NonDebuggable]
+    procedure GetMataDataValueFromResponseHeaders(): Dictionary of [Text, Text]
```
Comment (darjoo) [inline src/System Application/App/Azure Blob Services API/src/ABSOperationResponse.Codeunit.al]: Misspelling of name
```suggestion
    procedure GetMetaDataValueFromResponseHeaders(): Dictionary of [Text, Text]
```

File: src/System Application/App/Azure Blob Services API/src/ABSClientImpl.Codeunit.al [hunk 2]
```diff
@@ -554,6 +555,45 @@ codeunit 9051 "ABS Client Impl."
...
+    var
+        ABSOperationResponse: Codeunit "ABS Operation Response";
+        HTTPContent: HttpContent;
+        Operation: Enum "ABS Operation";
+    begin
+        ABSOperationPayload.SetOperation(Operation::SetBlobMetadata);
+        ABSOperationPayload.SetBlobName(BlobName);
+        AddMetaDataHeaders(MetaData);
+
+
```
Comment (darjoo) [inline src/System Application/App/Azure Blob Services API/src/ABSClientImpl.Codeunit.al]: Remove empty line

```suggestion
```

File: src/System Application/App/Azure Blob Services API/src/ABSOperationResponse.Codeunit.al [hunk 3]
```diff
@@ -93,6 +93,25 @@ codeunit 9050 "ABS Operation Response"
...
+        Headers := HttpResponseMessage.Headers();
+        HeaderKeys := Headers.Keys();
+        foreach HeaderKey in HeaderKeys do
+            if HeaderKey.StartsWith('x-ms-meta-') then
+                if Headers.GetValues(HeaderKey, Values) then
+                    MetaData.Add(DelStr(HeaderKey, 1, 10), Values[1]);
+        exit(MetaData);
+    end;
+
+
```
Comment (darjoo) [inline src/System Application/App/Azure Blob Services API/src/ABSOperationResponse.Codeunit.al]: Removal of empty line 
```suggestion
```

File: src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al [hunk 4]
```diff
@@ -1114,4 +1114,43 @@ codeunit 9053 "ABS Blob Client"
 
     var
         ABSClientImpl: Codeunit "ABS Client Impl.";
+
+    /// <summary>
+    /// The Get Blob MetaData operation gets user-defined MetaDatas for the specified blob as one or more key-value pairs.
```
Comment (grobyns) [inline src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al]:
```suggestion
    /// The GetBlobMetadata operation gets user-defined Metadata for the specified blob as one or more key-value pairs.
```

File: src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al [hunk 5]
```diff
@@ -1114,4 +1114,43 @@ codeunit 9053 "ABS Blob Client"
...
+    /// see: https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob-metadata?tabs=microsoft-entra-id
+    /// </summary>
+    /// <param name="BlobName">The name of the blob.</param>
+    /// <param name="MetaData">The result Dictionary of [Text, Text] with blob metadatas.</param>
+    /// <returns>An operation response object</returns>
+    procedure GetBlobMetaData(BlobName: Text; var MetaData: Dictionary of [Text, Text]): Codeunit "ABS Operation Response"
```
Comment (grobyns) [inline src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al]:
```suggestion
    procedure GetBlobMetadata(BlobName: Text; var Metadata: Dictionary of [Text, Text]): Codeunit "ABS Operation Response"
```

File: src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al [hunk 6]
```diff
@@ -1114,4 +1114,43 @@ codeunit 9053 "ABS Blob Client"
 
     var
         ABSClientImpl: Codeunit "ABS Client Impl.";
+
+    /// <summary>
+    /// The Get Blob MetaData operation gets user-defined MetaDatas for the specified blob as one or more key-value pairs.
+    /// see: https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob-metadata?tabs=microsoft-entra-id
+    /// </summary>
+    /// <param name="BlobName">The name of the blob.</param>
+    /// <param name="MetaData">The result Dictionary of [Text, Text] with blob metadatas.</param>
```
Comment (grobyns) [inline src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al]:
```suggestion
    /// <param name="Metadata">The result Dictionary of [Text, Text] with blob metadata.</param>
```

Comment (grobyns) [inline src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al]:
```suggestion
    /// <param name="MetaData">The result Dictionary of [Text, Text] with blob metadata.</param>
```

File: src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al [hunk 7]
```diff
@@ -1114,4 +1114,43 @@ codeunit 9053 "ABS Blob Client"
...
+    /// <returns>An operation response object</returns>
+    procedure GetBlobMetaData(BlobName: Text; var MetaData: Dictionary of [Text, Text]): Codeunit "ABS Operation Response"
+    var
+        OptionalParameters: Codeunit "ABS Optional Parameters";
+    begin
+        exit(ABSClientImpl.GetBlobMetaData(BlobName, MetaData, OptionalParameters))
```
Comment (grobyns) [inline src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al]:
```suggestion
        exit(ABSClientImpl.GetBlobMetadata(BlobName, Metadata, OptionalParameters))
```

File: src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al [hunk 8]
```diff
@@ -1114,4 +1114,43 @@ codeunit 9053 "ABS Blob Client"
...
+    end;
+
+    /// <summary>
+    /// The Get Blob MetaData operation gets user-defined MetaDatas for the specified blob as one or more key-value pairs.
```
Comment (grobyns) [inline src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al]: please apply the comment suggestions from above to all objects. 
Metadata -> 1 word, no plurals
procedure names should be used as such -> The GetBlobMetadata operation etc...

File: src/System Application/App/Azure Blob Services API/src/ABSClientImpl.Codeunit.al [hunk 9]
```diff
@@ -29,6 +29,7 @@ codeunit 9051 "ABS Client Impl."
         AppendBlockFromUrlOperationNotSuccessfulErr: Label 'Could not append block from URL %1 on %2.', Comment = '%1 = Source URI; %2 = Blob';
         GetPropertiesOperationNotSuccessfulErr: Label 'Could not get properties for %1.', Comment = '%1 = Blob';
         TagsOperationNotSuccessfulErr: Label 'Could not %1 %2 Tags.', Comment = '%1 = Get/Set, %2 = Service/Blob, ';
+        MetaDataOperationNotSuccessfulErr: Label 'Could not %1 %2 MetaData.', Comment = '%1 = Get/Set, %2 = Service/Blob, ';
```
Comment (grobyns) [inline src/System Application/App/Azure Blob Services API/src/ABSClientImpl.Codeunit.al]: These partial strings are really tough on translators and really should not be used. You only have 4 possible combinations here, so just spell them out, add them all. 

Yes, I can see ALL the other strings were done incorrectly as well... Someone should fix that. 

Also, you can see they created Labels for their hardcoded strings so you are missing GetLbl and SetLbl if you want to follow their pattern (which is very bad anyway).

File: src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al (hunk 8 above)
Comment (Indubhusan23) [reply src/System Application/App/Azure Blob Services API/src/ABSBlobClient.Codeunit.al]: Updated , Modified the Procedure name and other comments

File: src/System Application/App/Azure Blob Services API/src/ABSClientImpl.Codeunit.al (hunk 9 above)
Comment (Indubhusan23) [reply src/System Application/App/Azure Blob Services API/src/ABSClientImpl.Codeunit.al]: Updated,
Taken your feedback and made 2 different labels for get and set

File: src/System Application/Test/Azure Blob Services API/src/ABSBlobClientTest.Codeunit.al [hunk 10]
```diff
@@ -787,6 +787,52 @@ codeunit 132920 "ABS Blob Client Test"
         VerifyContainerContentType(ABSContainerContent, CopyStr(ABSTestLibrary.GetSampleResponseFileName(), 1, MaxStrLen(ABSContainerContent.Name)), Enum::"ABS Blob Resource Type"::File);
     end;
 
+    [Test] // Failing if ran against azurite, as BLOB MetaData are not supported there
+    procedure GetBlockBlobMetaDataTest()
```
Comment (JesperSchulz) [inline src/System Application/Test/Azure Blob Services API/src/ABSBlobClientTest.Codeunit.al:791]: Given this test is failing in the CI pipeline, you must disable it here: https://github.com/microsoft/BCApps/blob/main/src/System%20Application/Test/DisabledTests/ABSBlobClientTest.json

**Comment by JesperSchulz** [comment]
@Indubhusan23, once the test is disabled and the merge conflict is resolved, we'll take another look at this one and get it ready for merging.

**Comment by Indubhusan23** [comment]
> @Indubhusan23, once the test is disabled and the merge conflict is resolved, we'll take another look at this one and get it ready for merging.
@JesperSchulz disabled the unit test and resolved the merge conflicts

File: src/System Application/Test/Azure Blob Services API/src/ABSBlobClientTest.Codeunit.al (hunk 10 above)
Comment (Indubhusan23) [reply src/System Application/Test/Azure Blob Services API/src/ABSBlobClientTest.Codeunit.al:791]: done

**Comment by Indubhusan23** [comment]
@JesperSchulz can you have a look on this PR as you requested i have disabled the unit test and the solved the merge conflict.

**Review by JesperSchulz — APPROVED** [review]
Looks good to me!

File: src/System Application/App/Azure Blob Services API/src/ABSOperationResponse.Codeunit.al [hunk 11]
```diff
@@ -93,6 +93,24 @@ codeunit 9050 "ABS Operation Response"
...
     end;
 
+    [NonDebuggable]
+    procedure GetMetadataValueFromResponseHeaders(): Dictionary of [Text, Text]
+    var
+        Metadata: Dictionary of [Text, Text];
+        Headers: HttpHeaders;
+        HeaderKeys: List of [Text];
+        HeaderKey: Text;
+        Values: array[100] of Text;
```
Comment (darjoo) [inline src/System Application/App/Azure Blob Services API/src/ABSOperationResponse.Codeunit.al:103]: Is there a reason this needs to be 100?

**Review by darjoo — APPROVED** [review]

File: src/System Application/App/Azure Blob Services API/src/ABSOperationResponse.Codeunit.al (hunk 11 above)
Comment (Indubhusan23) [reply src/System Application/App/Azure Blob Services API/src/ABSOperationResponse.Codeunit.al:103]: No there is No particular reason , as the array need to be provided with size and for safer side i have given the number 100

**Pull request merged**